cmd/internal.go

```diff
@@ -208,3 +229,64 @@ func PatchAsset() *cli.Command {
 		},
 	}
 }
+
+// ParseLineage analyzes the column lineage for a given asset within a pipeline.
+// It traces column relationships between the asset and its upstream dependencies.
+func ParseLineage(pipe *pipeline.Pipeline, asset *pipeline.Asset) error {
+	parser, err := sqlparser.NewSQLParser()
+	if err != nil {
+		return fmt.Errorf("failed to create SQL parser: %w", err)
+	}
+
+	if err := parser.Start(); err != nil {
+		return fmt.Errorf("failed to start SQL parser: %w", err)
+	}
+
+	columnMetadata := make(sqlparser.Schema)
+	for _, upstream := range asset.Upstreams {
+		upstreamAsset := pipe.GetAssetByName(upstream.Value)
+		if upstreamAsset == nil {
+			return fmt.Errorf("upstream asset not found: %s", upstream.Value)
+		}
+		columnMetadata[upstreamAsset.Name] = makeColumnMap(upstreamAsset.Columns)
+	}
+
+	lineage, err := parser.ColumnLineage(asset.ExecutableFile.Content, "", columnMetadata)
+	if err != nil {
+		return fmt.Errorf("failed to parse column lineage: %w", err)
+	}
+
+	for _, lineageCol := range lineage.Columns {
+		for _, upstream := range lineageCol.Upstream {
+			upstreamAsset := pipe.GetAssetByName(upstream.Table)
+			if upstreamAsset == nil {
+				return fmt.Errorf("upstream asset not found: %s", upstream.Table)
+			}
+
+			upstreamCol := upstreamAsset.GetColumnWithName(upstream.Column)
+			if upstreamCol == nil {
+				return fmt.Errorf("column %s not found in asset %s", upstream.Column, upstream.Table)
+			}
```

Comment:
same goes for this, maybe the upstream did not define the column, we shouldn't error here.